### PR TITLE
PWGHF: addition of D0 ML response class

### PR DESCRIPTION
**Pull request by atavirag**
Following @AlexBigO's ML response class for Dplus, I made an equivalent class for D0 mesons. I added the variables currently available in treeCreatorD0ToKPi.cxx.

### CODE REVIEW

**Comment by AlexBigO** [comment]
Hello @atavirag , I don't have any comments on the header but I have a few regarding the selector, otherwise it looks good to me :)
1. You should define a configurable array containing the names of the wanted input features: as done [here](https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/candidateSelectorDplusToPiKPi.cxx#L68)
2. You need to cache the input indices to translate your configurables input features names into integers (via the `enum` defined in the header): you can use [this line](https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/candidateSelectorDplusToPiKPi.cxx#L122)
3. You should change the way `aod::HfMlD0` is filled when a candidate does not pass preselections. This is a bug spotted by Fabio that we corrected in a recent PR for Dplus. The idea is to define an empty vector ([see here](https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/candidateSelectorDplusToPiKPi.cxx#L77)) and then feed this vector to `hfMlD0Candidate` when preselections (Skim, Topol, Pid) are not fulfilled by the candidate (as done [here](https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/candidateSelectorDplusToPiKPi.cxx#L209) for instance).

**Comment by atavirag** [comment]
Hi @AlexBigO, I have just added your comments. Thank you!

File: PWGHF/TableProducer/candidateSelectorD0.cxx
Comment (AlexBigO) [inline PWGHF/TableProducer/candidateSelectorD0.cxx:121]: Everything is okay on my side. Just a minor comment, now you don't need this line anymore as the `outputMl` vector (defined above `init`) is going to be passed by reference in `isSelectedMl`.

File: PWGHF/Core/HfMlResponseD0ToKPi.h
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseD0ToKPi.h:72]: These variables are not needed given that they should not be useful for BDTs

File: PWGHF/Core/HfMlResponseD0ToKPi.h
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseD0ToKPi.h:74]: I would remove this one too

File: PWGHF/Core/HfMlResponseD0ToKPi.h
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseD0ToKPi.h:81]: no need to put the squared momenta (the pt is sufficient)

File: PWGHF/Core/HfMlResponseD0ToKPi.h
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseD0ToKPi.h:89]: I would say that the single components of the daughter momenta are not needed

File: PWGHF/Core/HfMlResponseD0ToKPi.h
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseD0ToKPi.h:93]: this is also not needed

File: PWGHF/Core/HfMlResponseD0ToKPi.h
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseD0ToKPi.h:102]: The mass should never be used as feature, so I would remove it

File: PWGHF/Core/HfMlResponseD0ToKPi.h
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseD0ToKPi.h:107]: The momenta of the candidate also should not be used

File: PWGHF/Core/HfMlResponseD0ToKPi.h
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseD0ToKPi.h:114]: Same for these variables

File: PWGHF/Core/HfMlResponseD0ToKPi.h
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseD0ToKPi.h:101]: In addition to the single Nsigma TPC and TOF variables, I would also add the combined one (see e.g. https://github.com/AliceO2Group/O2Physics/blob/1204d8f12614b5601381c70138bcdee4224ccf5b/PWGHF/Core/HfMlResponseDplusToPiKPi.h#L78-L83)

File: PWGHF/Core/HfMlResponseD0ToKPi.h
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseD0ToKPi.h:129]:
```suggestion
  /// \param candidate is the D0 candidate
```

File: PWGHF/Core/HfMlResponseD0ToKPi.h
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseD0ToKPi.h:136]: Differently from the Dplus, in the case of the D0 we should handle the two possible mass hypotheses (KPi or PiK). This means that we should have the flexibility to fill the vector of variables with one hypothesis or the other for all the variables that depend on this mass hypothesis, such as the cost*, see https://github.com/AliceO2Group/O2Physics/blob/1204d8f12614b5601381c70138bcdee4224ccf5b/PWGHF/Core/HfHelper.h#L73-L83
 My proposal would be to add an argument here that specifies whether the vector should be filled for the hypothesis `D0toPiK` (first daughter as a pion, second one as a kaon) or the hypothesis `D0bartoKPi` (first daughter as a kaon, second one as a pion)

File: PWGHF/TableProducer/candidateSelectorD0.cxx
Comment (fgrosa) [inline PWGHF/TableProducer/candidateSelectorD0.cxx:363]: In the hard-coded version we did not have variables that depend on the mass hypothesis. However, now that we have the flexibility to modify the variables via the configurable, we should differentiate the two. In particular, if the candidate is preselected as a `D0toPiK` (`statusD0`) we should get the BDT output scores corresponding to that mass hypothesis, while if selected as `D0bartoKPi` (`statusD0bar`) the other ones. This means that sometimes the `isSelectedMl` method will be called for one of the two, sometimes for both. We should then decide a convention how to store them in the `hfMlD0Candidate` table (@fcatalan92 any idea?)

**Review by fgrosa — CHANGES_REQUESTED** [review]
Hi @atavirag thanks a lot for the implementation! I added few comments regarding variables to remove or add and then the other missing thing to address (not present for the D+) is the way to handle the two possible mass hypotheses for the D0/D0bar candidates.

File: PWGHF/TableProducer/candidateSelectorD0.cxx
Comment (fgrosa) [reply PWGHF/TableProducer/candidateSelectorD0.cxx:363]: Yes this might indeed be a possibility. In this case we should decide whether to fill the columns with empty vectors in case there are no BDT output scores for one of the hypothesis (based on the `statusD0` or `statusD0bar` before the BDT application) or to always test the BDT for both hypotheses and fill it.

**Comment by atavirag** [comment]
Hi @fgrosa, I just made a commit changing the variables. I also added the column `hf_sel_candidate_d0::MlProbD0bar` and, at least for now, made it so that it always tests the BDT output for both mass hypotheses. Please let me know if it's okay or if you'd prefer the other option to fill the table. Thanks!

File: PWGHF/TableProducer/candidateSelectorD0.cxx
Comment (fgrosa) [inline PWGHF/TableProducer/candidateSelectorD0.cxx:77]: There is no need to have these variables duplicated
```suggestion
  std::vector<float> outputMlD0 = {};
  std::vector<float> outputMlD0bar = {};
```

File: PWGHF/TableProducer/candidateSelectorD0.cxx
Comment (fgrosa) [inline PWGHF/TableProducer/candidateSelectorD0.cxx:285]:
```suggestion
          hfMlD0Candidate(outputMlD0, outputMlD0bar);
```

File: PWGHF/TableProducer/candidateSelectorD0.cxx
Comment (fgrosa) [inline PWGHF/TableProducer/candidateSelectorD0.cxx:302]:
```suggestion
          hfMlD0Candidate(outputMlD0, outputMlD0bar);
```

File: PWGHF/TableProducer/candidateSelectorD0.cxx
Comment (fgrosa) [inline PWGHF/TableProducer/candidateSelectorD0.cxx:350]:
```suggestion
          hfMlD0Candidate(outputMlD0, outputMlD0bar);
```

File: PWGHF/TableProducer/candidateSelectorD0.cxx
Comment (fgrosa) [inline PWGHF/TableProducer/candidateSelectorD0.cxx:374]:
```suggestion
        if (!isSelectedMlD0) {
          statusD0 = 0;
        }
        if (!isSelectedMlD0bar) {
          statusD0bar = 0;
        }
```

File: PWGHF/TableProducer/candidateSelectorD0.cxx
Comment (fgrosa) [inline PWGHF/TableProducer/candidateSelectorD0.cxx:271]:
```suggestion
          hfMlD0Candidate(outputMlD0, outputMlD0bar);
```

Comment (fgrosa) [inline PWGHF/TableProducer/candidateSelectorD0.cxx:271]: Before this, e.g. in line `267` you should add
```
outputMlD0.clear();
outputMlD0bar.clear();
```
I cannot put it as a suggestion because that line is not modifed in your PR.

**Review by fgrosa — CHANGES_REQUESTED** [review]
Hi @atavirag thanks! I still have a couple of suggestions, then I would wait for @fcatalan92's feedback on the table with the two vectors and the decision whether to apply anyway always the ML model to both mass hypotheses (as in your implementation) or instead whether to do it only for the hypothesis which is preselected (maybe better).

**Comment by fcatalan92** [comment]
Hi @atavirag and @fgrosa, I would apply the model only to candidates that pass the pre-selections, so that we do not perform the model computation if not needed. I would also leave the vector empty in case the model is not applied, this prevents to use a wrong information later on, since if the vector is accessed we should get a crash (the correct way should be to check the flag of the ML selection before accessing the ML output).

File: PWGHF/TableProducer/candidateSelectorD0.cxx
Comment (fcatalan92) [inline PWGHF/TableProducer/candidateSelectorD0.cxx:387]: You should always fill the table, otherwise it will be not joinable with those of the candidates and selection flag.

**Review by fcatalan92 — CHANGES_REQUESTED** [review]
Hi @atavirag, thanks for the updates. I added a comment on the code.
Moreover, I have another remark: at the end of the part related to the ML application, you should clear the vectors, otherwise the ML output will be wrongly filled, e.g., also for candidates that do not pass the pre-selections.

Before exiting the `if`  you can do

```C++
    outputMlD0.clear();
    outputMlD0bar.clear();
```

**Comment by atavirag** [comment]
> Hi @atavirag, thanks for the updates. I added a comment on the code. Moreover, I have another remark: at the end of the part related to the ML application, you should clear the vectors, otherwise the ML output will be wrongly filled, e.g., also for candidates that do not pass the pre-selections.
> 
> Before exiting the `if` you can do
> 
> ```c++
>     outputMlD0.clear();
>     outputMlD0bar.clear();
> ```

The vectors are currently being cleared at the beginning of the for loop, in lines 266 and 267. That is, before they get the chance to get filled again. Would it be okay to leave it like that?

**Review by fcatalan92 — APPROVED** [review]

**Comment by vkucera** [comment]
This PR introduced a bug which prevents the workflow from running.
Fixed in https://github.com/AliceO2Group/O2Physics/pull/3872
@atavirag Please always test your changes locally before making a PR.

**Comment by atavirag** [comment]
> This PR introduced a bug which prevents the workflow from running. Fixed in #3872 @atavirag Please always test your changes locally before making a PR. Also this new subscription should be mentioned in the PR description as it is not directly related to the ML response.

I am sorry, thank you for the fix. I will add the subscription to the description.

**Pull request merged**
